### PR TITLE
Coords opts

### PR DESCRIPTION
**Pull request by d-v-b**
This addition adds the ability to get the set of (x,y) points inside a polygon drawn in an instance of `lightning.image` 

To test this new functionality, draw a region on an image and call the following:

```
# get set of pixels from image
coords = viz.get_coords(as_points=True)[0]
# build an array with the same size in (x,y) as the image data
tmp = np.zeros(dims[0:2])

# for each pixel, color the image at that point.
for pix in coords:
        tmp[pix[0],pix[1]] = 1

# the resulting binary image should match the shape and position of the drawn polygon
plt.imshow(tmp)

```

### CODE REVIEW

File: lightning/types/utils.py [hunk 1]
```diff
@@ -206,4 +206,27 @@ def list_to_regions(reg):
...
+    path = Path(bounds)
+
+    grid = meshgrid(range(bMin[0],bMax[0]+1),range(bMin[1],bMax[1]+1))
+
+    gridFlat = zip(grid[0].ravel(),grid[1].ravel())
+
+    inside = path.contains_points(gridFlat).reshape(grid[0].shape).astype('int')
+    inside = where(inside)
+    inside = vstack([inside[0],inside[1]]).T + bMin[-1::-1]    
+    inside.tolist()
```
Comment (freeman-lab) [inline lightning/types/utils.py]: `tolist()` is not in place, make sure to call it on the object that will be returned.

File: lightning/types/utils.py [hunk 2]
```diff
@@ -206,4 +206,27 @@ def list_to_regions(reg):
...
-        return reg
\ No newline at end of file
+        return reg
+
+def get_points_from_polygon(coords):
+    """
+    Given a list of pairs of points which define a polygon, return a list of points interior to the polygon
+    """
+    
+    bounds = array(coords).astype('int')
```
Comment (freeman-lab) [inline lightning/types/utils.py]: Replace `array` with `asarray`

File: lightning/types/utils.py [hunk 3]
```diff
@@ -206,4 +206,27 @@ def list_to_regions(reg):
...
+
+    bMax = bounds.max(0)
+    bMin = bounds.min(0)
+
+    path = Path(bounds)
+
+    grid = meshgrid(range(bMin[0],bMax[0]+1),range(bMin[1],bMax[1]+1))
```
Comment (freeman-lab) [inline lightning/types/utils.py]: Here and elsewhere put spaces after commas within function argument lists

**Comment by freeman-lab** [comment]
@d-v-b this looks great! I think it would be nice to add one more output type, a mask for each region, of 0s and 1s with 1s where the region was defined. As we'll then have three options, I would suggest the following change to the name / function definition:

```
def get_regions(self, return_type='bounds', dims=None):
```

Where the three options are `bounds`, `points`, and `mask`. Then the core of the method would look like:

```
    coords = user_data['coords']
    if return_type == 'bounds':
        return coords
    if return_type == 'points':
        return [get_points_from_polygon(x) for x in coords]
    else if return_type == 'mask':
        if not dims:
            raise Exception('Must provide dimensions to return masks')
        # call function to generate a mask on each region
    else:
        raise Exception('Option %s not supported' % return_type)
```

Want to give this a try?

**Comment by d-v-b** [comment]
@freeman-lab yep, I can work those suggestions in. Expect something along those lines in a few hours

**Comment by d-v-b** [comment]
A list of boundary points for each polygon are returned when `get_coords` is called like this:
 `bounds = viz.get_coords()`

Given `dims`, the dimensions of the image containing the polygons, a list of binary masks formed by each polygon is returned when `get_coords` is called like this:
`mask = viz.get_coords(return_type='mask',dims=dims)`

A list of lists of points internal to each polygon is returned when `get_coords` is called like this:
`pts = viz.get_coords(return_type='points',dims=dims)` 

The argument `dims` is required for both inner points and masks because the masks are generated by calling the function `bounds_to_mask(bounds,mask)` on the bounds, and the points are subsequently generated by calling `mask_to_points(mask)`

File: lightning/types/utils.py [hunk 4]
```diff
@@ -206,4 +206,35 @@ def list_to_regions(reg):
...
+    grid = meshgrid(range(dims[1]), range(dims[0]))
+    grid_flat = zip(grid[0].ravel(), grid[1].ravel())
+
+    mask = path.contains_points(grid_flat).reshape(dims[0:2]).astype('int')
+
+    return mask
+
+
+def mask_to_points(mask):
+
```
Comment (freeman-lab) [inline lightning/types/utils.py]: Remove line break

File: lightning/types/utils.py [hunk 5]
```diff
@@ -206,4 +206,35 @@ def list_to_regions(reg):
         checkthree = all(map(lambda x: len(x) == 3, reg))
         if not (checktwo or checkthree):
             raise Exception("All region names must be two letters (for US) or three letters (for world)")
-        return reg
\ No newline at end of file
+        return reg
+
+
+def polygon_to_mask(coords, dims):
+
```
Comment (freeman-lab) [inline lightning/types/utils.py]: Remove line break

File: lightning/types/images.py [hunk 6]
```diff
@@ -27,12 +27,34 @@ def clean(imagedata):
         outdict = [array_to_im(imagedata)]
 
         return {'images': outdict}
+    
+    def get_coords(self, return_type='bounds', dims=None):
+        """
+        Get data from polygons drawn on image.
+
+        :param return_type:
```
Comment (freeman-lab) [inline lightning/types/images.py]: Follow arg formatting of other functions, it should look like:

```
Parameters
----------
return_type : str
    <Explanation of parameter and its options>
dims : tuple, list, or array, optional, default=None
    <Explanation of parameters>
```

**Comment by freeman-lab** [comment]
@d-v-b looking very close! Made a couple formatting suggestions, only other thing is that I'd rather use the `get_points_from_polygon` from before instead of `mask_to_points(polygon_to_mask())`, becuase the earlier one didn't require knowing the `dims`. Normally I'd prefer the modularity and consistency of your approach, but I think the ability to get points without providing the dimensions is actually nicer to have, and you need an extra function in either case.

**Comment by d-v-b** [comment]
@freeman-lab now `get_coords(return_type='points')` does not require `dims` as an argument, thanks to the addition of a function `polygon_to_points`

for consistency in naming, I also re-named the mask-generating function to `polygon_to_mask`.

**Comment by freeman-lab** [comment]
@d-v-b Terrific, LGTM, really nice work! Merging it in.

**Pull request merged**